### ISSUE TITLE
Add conditional mapping

### ISSUE THREAD

**Issue by embray** [astropy/astropy#3393]
I still need to think some on exactly what it would look like, but a special type of [mapping model](http://docs.astropy.org/en/latest/modeling/compound-models.html#advanced-mappings) that select between different models depending on a conditional expression would also be useful.  This would make it mostly straightforward to define piecewise functions in the compound model framework.

**Comment by mdboom** [comment]
I think some of the instruments in JWST will need this...

**Comment by nden** [comment]
This is implemented here
https://github.com/spacetelescope/gwcs/blob/master/gwcs/selector.py

**Comment by mhvk** [comment]
A risk here is that the selection/mapping model has to become fully aware of dimensions of inputs, etc. Possibly, the logic for what to select should not reside in the mapping/selecting model, but in the models underneath, which could, say, return `None` or `False` or `NaN` when input is outside of their region of validity. Then, one would only need a joining model that selects the one (and only one) valid set of results. Or maybe more simply there would be a validation wrapper for models that does the above.

But I do worry that if one were to proceed along these paths before too long a whole new programming language would have been defined...

**Comment by embray** [comment]
I'm not really sure I follow, but more constraints that can be set on individual models is also a good idea.  Right now the constraints that exist are *only* used by fitters but there has been an open issue to make them apply to general evaluation.  Also I don't think there is currently a way to set constraints on output (though in most cases a constraint on input can do that too).

**Comment by nden** [comment]
@mhvk You are right, there's assumption that the selector model knows the input dimensions. If the region of validity for each child model is stored within that model, isn't it true that then child models would need to know about the input dimensions? In other words, if model for region 3 is valid over pixel space [128 : 230, 30:2000] what is the zero point for these coordinates? Presumably it's the coordinate system of the inputs which makes individual models aware of the inputs.

But I do think models need to have region of validity (I'm calling it shortly "domain"). 
And there may be different use cases for both approaches.

**Comment by mhvk** [comment]
@nden - yes, makes sense that sometimes you do not want the underlying model to have to care about all the details like number of pixels and so on.  Rather complicated...

**Comment by embray** [comment]
I think we might be talking at cross purposes here because I don't think that's really what I had in mind.
